File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/src/canvas.dart
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ui/ui.dart' as ui;

import 'paint.dart';
import 'path.dart';
import 'picture.dart';
import 'raw/raw_canvas.dart';
import 'raw/raw_memory.dart';
import 'raw/raw_picture.dart';

class SkwasmCanvas implements ui.Canvas {
  factory SkwasmCanvas(SkwasmPictureRecorder recorder, [ui.Rect cullRect = ui.Rect.largest]) =>
    SkwasmCanvas.fromHandle(withStackScope((StackScope s) =>
      pictureRecorderBeginRecording(recorder.handle, s.convertRect(cullRect))
    ));

  SkwasmCanvas.fromHandle(this._handle);
  CanvasHandle _handle;

  void delete() {
    canvasDestroy(_handle);
  }

  @override
  void save() {
    canvasSave(_handle);
  }

  @override
  void saveLayer(ui.Rect? bounds, ui.Paint uiPaint) {
    assert(uiPaint is SkwasmPaint);
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    if (bounds != null) {
      withStackScope((StackScope s) {
        canvasSaveLayer(_handle, s.convertRect(bounds), paint.handle);
      });
    } else {
      canvasSaveLayer(_handle, nullptr, paint.handle);
    }
  }

  @override
  void restore() {
    canvasRestore(_handle);
  }

  @override
  int getSaveCount() => canvasGetSaveCount(_handle);

  @override
  void translate(double dx, double dy) => canvasTranslate(_handle, dx, dy);

  @override
  void scale(double sx, [double? sy]) => canvasScale(_handle, sx, sy ?? sx);

  @override
  void rotate(double radians) => canvasRotate(_handle, ui.toDegrees(radians));

  @override
  void skew(double sx, double sy) => canvasSkew(_handle, sx, sy);

  @override
  void transform(Float64List matrix4) {
    withStackScope((StackScope s) {
      canvasTransform(_handle, s.convertMatrix4toSkM44(matrix4));
    });
  }

  @override
  void clipRect(ui.Rect rect,
      {ui.ClipOp clipOp = ui.ClipOp.intersect, bool doAntiAlias = true}) {
    withStackScope((StackScope s) {
      canvasClipRect(_handle, s.convertRect(rect), clipOp.index, doAntiAlias);
    });
  }

  @override
  void clipRRect(ui.RRect rrect, {bool doAntiAlias = true}) {
    withStackScope((StackScope s) {
      canvasClipRRect(_handle, s.convertRRect(rrect), doAntiAlias);
    });
  }

  @override
  void clipPath(ui.Path uiPath, {bool doAntiAlias = true}) {
    assert(uiPath is SkwasmPath);
    final SkwasmPath path = uiPath as SkwasmPath;
    canvasClipPath(_handle, path.handle, doAntiAlias);
  }

  @override
  void drawColor(ui.Color color, ui.BlendMode blendMode) => 
    canvasDrawColor(_handle, color.value, blendMode.index);

  @override
  void drawLine(ui.Offset p1, ui.Offset p2, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    canvasDrawLine(_handle, p1.dx, p1.dy, p2.dx, p2.dy, paint.handle);
  }

  @override
  void drawPaint(ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    canvasDrawPaint(_handle, paint.handle);
  }

  @override
  void drawRect(ui.Rect rect, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    withStackScope((StackScope s) {
      canvasDrawRect(_handle, s.convertRect(rect), paint.handle);
    });
  }

  @override
  void drawRRect(ui.RRect rrect, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    withStackScope((StackScope s) {
      canvasDrawRRect(_handle, s.convertRRect(rrect), paint.handle);
    });
  }

  @override
  void drawDRRect(ui.RRect outer, ui.RRect inner, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    withStackScope((StackScope s) {
      canvasDrawDRRect(
          _handle, s.convertRRect(outer), s.convertRRect(inner), paint.handle);
    });
  }

  @override
  void drawOval(ui.Rect rect, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    withStackScope((StackScope s) {
      canvasDrawOval(_handle, s.convertRect(rect), paint.handle);
    });
  }

  @override
  void drawCircle(ui.Offset center, double radius, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    canvasDrawCircle(_handle, center.dx, center.dy, radius, paint.handle);
  }

  @override
  void drawArc(ui.Rect rect, double startAngle, double sweepAngle, bool useCenter,
      ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    withStackScope((StackScope s) {
      canvasDrawArc(
          _handle,
          s.convertRect(rect),
          ui.toDegrees(startAngle),
          ui.toDegrees(sweepAngle),
          useCenter,
          paint.handle);
    });
  }

  @override
  void drawPath(ui.Path uiPath, ui.Paint uiPaint) {
    final SkwasmPaint paint = uiPaint as SkwasmPaint;
    final SkwasmPath path = uiPath as SkwasmPath;
    canvasDrawPath(_handle, path.handle, paint.handle);
  }

  @override
  void drawImage(ui.Image uiImage, ui.Offset offset, ui.Paint uiPaint) {
    throw UnimplementedError();
  }

  @override
  void drawImageRect(ui.Image uiImage, ui.Rect src, ui.Rect dst, ui.Paint uiPaint) {
    throw UnimplementedError();
  }

  @override
  void drawImageNine(ui.Image uiImage, ui.Rect center, ui.Rect dst, ui.Paint uiPaint) {
    throw UnimplementedError();
  }

  @override
  void drawPicture(ui.Picture picture) {
    canvasDrawPicture(_handle, (picture as SkwasmPicture).handle);
  }

  @override
  void drawParagraph(ui.Paragraph uiParagraph, ui.Offset offset) {
    throw UnimplementedError();
  }

  @override
  void drawPoints(ui.PointMode pointMode, List<ui.Offset> points, ui.Paint paint) {
    throw UnimplementedError();
  }

  @override
  void drawRawPoints(ui.PointMode pointMode, Float32List points, ui.Paint paint) {
    throw UnimplementedError();
  }

  @override
  void drawVertices(ui.Vertices vertices, ui.BlendMode blendMode, ui.Paint paint) {
    throw UnimplementedError();
  }

  @override
  void drawAtlas(
    ui.Image atlas,
    List<ui.RSTransform> transforms,
    List<ui.Rect> rects,
    List<ui.Color>? colors,
    ui.BlendMode? blendMode,
    ui.Rect? cullRect,
    ui.Paint paint,
  ) {
    throw UnimplementedError();
  }

  @override
  void drawRawAtlas(
    ui.Image atlas,
    Float32List rstTransforms,
    Float32List rects,
    Int32List? colors,
    ui.BlendMode? blendMode,
    ui.Rect? cullRect,
    ui.Paint paint,
  ) {
    throw UnimplementedError();
  }

  @override
  void drawShadow(
    ui.Path path,
    ui.Color color,
    double elevation,
    bool transparentOccluder,
  ) {
    throw UnimplementedError();
  }
  
  @override
  ui.Rect getDestinationClipBounds() {
    // TODO: implement getDestinationClipBounds
    throw UnimplementedError();
  }
  
  @override
  ui.Rect getLocalClipBounds() {
    // TODO: implement getLocalClipBounds
    throw UnimplementedError();
  }
  
  @override
  Float64List getTransform() {
    // TODO: implement getTransform
    throw UnimplementedError();
  }
  
  @override
  void restoreToCount(int count) {
    // TODO: implement restoreToCount
  }
}

File: lib/web_ui/skwasm/canvas.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "helpers.h"
#include "render_strategy.h"
#include "wrappers.h"

using namespace Skwasm;

SKWASM_EXPORT void canvas_saveLayer(Canvas* canvas,
                                    Rect* rect,
                                    Paint* paint,
                                    ImageFilter* backdrop) {
  canvas->saveLayer(Canvas::SaveLayerRec(rect, paint, backdrop, 0));
}

SKWASM_EXPORT void canvas_save(Canvas* canvas) {
  canvas->save();
}

SKWASM_EXPORT void canvas_restore(Canvas* canvas) {
  canvas->restore();
}

SKWASM_EXPORT void canvas_restoreToCount(Canvas* canvas, int count) {
  canvas->restoreToCount(count);
}

SKWASM_EXPORT int canvas_getSaveCount(Canvas* canvas) {
  return canvas->getSaveCount();
}

SKWASM_EXPORT void canvas_translate(Canvas* canvas, Scalar dx, Scalar dy) {
  canvas->translate(dx, dy);
}

SKWASM_EXPORT void canvas_scale(Canvas* canvas, Scalar sx, Scalar sy) {
  canvas->scale(sx, sy);
}

SKWASM_EXPORT void canvas_rotate(Canvas* canvas, Scalar degrees) {
  canvas->rotate(degrees);
}

SKWASM_EXPORT void canvas_skew(Canvas* canvas, Scalar sx, Scalar sy) {
  canvas->skew(sx, sy);
}

SKWASM_EXPORT void canvas_transform(Canvas* canvas, const Matrix44* matrix44) {
  canvas->concat(*matrix44);
}

SKWASM_EXPORT void canvas_clipRect(Canvas* canvas,
                                   const Rect* rect,
                                   ClipOp op,
                                   bool antialias) {
  canvas->clipRect(*rect, op, antialias);
}

SKWASM_EXPORT void canvas_clipRRect(Canvas* canvas,
                                    const Scalar* rrectValues,
                                    bool antialias) {
  canvas->clipRRect(createRRect(rrectValues), antialias);
}

SKWASM_EXPORT void canvas_clipPath(Canvas* canvas, Path* path, bool antialias) {
  canvas->clipPath(*path, antialias);
}

SKWASM_EXPORT void canvas_drawColor(Canvas* canvas,
                                    Color color,
                                    BlendMode blendMode) {
  canvas->drawColor(color, blendMode);
}

SKWASM_EXPORT void canvas_drawLine(Canvas* canvas,
                                   Scalar x1,
                                   Scalar y1,
                                   Scalar x2,
                                   Scalar y2,
                                   Paint* paint) {
  canvas->drawLine(x1, y1, x2, y2, *paint);
}

SKWASM_EXPORT void canvas_drawPaint(Canvas* canvas, Paint* paint) {
  canvas->drawPaint(*paint);
}

SKWASM_EXPORT void canvas_drawRect(Canvas* canvas, Rect* rect, Paint* paint) {
  canvas->drawRect(*rect, *paint);
}

SKWASM_EXPORT void canvas_drawRRect(Canvas* canvas,
                                    const Scalar* rrectValues,
                                    Paint* paint) {
  canvas->drawRRect(createRRect(rrectValues), *paint);
}

SKWASM_EXPORT void canvas_drawDRRect(Canvas* canvas,
                                     const Scalar* outerRrectValues,
                                     const Scalar* innerRrectValues,
                                     Paint* paint) {
  canvas->drawDRRect(createRRect(outerRrectValues),
                     createRRect(innerRrectValues), *paint);
}

SKWASM_EXPORT void canvas_drawOval(Canvas* canvas,
                                   const Rect* rect,
                                   Paint* paint) {
  canvas->drawOval(*rect, *paint);
}

SKWASM_EXPORT void canvas_drawCircle(Canvas* canvas,
                                     Scalar x,
                                     Scalar y,
                                     Scalar radius,
                                     Paint* paint) {
  canvas->drawCircle(x, y, radius, *paint);
}

SKWASM_EXPORT void canvas_drawArc(Canvas* canvas,
                                  const Rect* rect,
                                  Scalar startAngleDegrees,
                                  Scalar sweepAngleDegrees,
                                  bool useCenter,
                                  Paint* paint) {
  canvas->drawArc(*rect, startAngleDegrees, sweepAngleDegrees, useCenter,
                  *paint);
}

SKWASM_EXPORT void canvas_drawPath(Canvas* canvas, Path* path, Paint* paint) {
  canvas->drawPath(*path, *paint);
}

SKWASM_EXPORT void canvas_drawShadow(Canvas* canvas,
                                     Path* path,
                                     Scalar elevation,
                                     Scalar devicePixelRatio,
                                     Color color,
                                     bool transparentOccluder) {
  drawShadowOnCanvas(canvas, path, elevation, devicePixelRatio, color,
                     transparentOccluder);
}

SKWASM_EXPORT void canvas_drawParagraph(Canvas* canvas,
                                        Paragraph* paragraph,
                                        Scalar x,
                                        Scalar y) {
  drawParagraphOnCanvas(canvas, paragraph, x, y);
}

SKWASM_EXPORT void canvas_drawPicture(Canvas* canvas, Picture* picture) {
  canvas->drawPicture(picture);
}

SKWASM_EXPORT void canvas_drawImage(Canvas* canvas,
                                    Image* image,
                                    Scalar offsetX,
                                    Scalar offsetY,
                                    Paint* paint,
                                    FilterQuality quality) {
  canvas->drawImage(image, offsetX, offsetY, samplingOptionsForQuality(quality),
                    paint);
}

SKWASM_EXPORT void canvas_drawImageRect(Canvas* canvas,
                                        Image* image,
                                        Rect* sourceRect,
                                        Rect* destRect,
                                        Paint* paint,
                                        FilterQuality quality) {
  canvas->drawImageRect(image, *sourceRect, *destRect,
                        samplingOptionsForQuality(quality), paint,
                        Canvas::kStrict_SrcRectConstraint);
}

SKWASM_EXPORT void canvas_drawImageNine(Canvas* canvas,
                                        Image* image,
                                        IRect* centerRect,
                                        Rect* destinationRect,
                                        Paint* paint,
                                        FilterQuality quality) {
  canvas->drawImageNine(image, *centerRect, *destinationRect,
                        filterModeForQuality(quality), paint);
}

SKWASM_EXPORT void canvas_drawVertices(Canvas* canvas,
                                       Vertices* vertices,
                                       BlendMode mode,
                                       Paint* paint) {
  canvas->drawVertices(vertices, mode, *paint);
}

SKWASM_EXPORT void canvas_drawPoints(Canvas* canvas,
                                     Canvas::PointMode mode,
                                     Point* points,
                                     int pointCount,
                                     Paint* paint) {
  canvas->drawPoints(mode, pointCount, points, *paint);
}

SKWASM_EXPORT void canvas_drawAtlas(Canvas* canvas,
                                    Image* atlas,
                                    RSXform* transforms,
                                    Rect* rects,
                                    Color* colors,
                                    int spriteCount,
                                    BlendMode mode,
                                    Rect* cullRect,
                                    Paint* paint) {
  canvas->drawAtlas(atlas, transforms, rects, colors, spriteCount, mode,
                    SamplingOptions{FilterMode::kLinear, MipmapMode::kNone},
                    cullRect, paint);
}

SKWASM_EXPORT void canvas_getTransform(Canvas* canvas, Matrix44* outTransform) {
  *outTransform = canvas->getLocalToDevice();
}

SKWASM_EXPORT void canvas_getLocalClipBounds(Canvas* canvas, Rect* outRect) {
  *outRect = canvas->getLocalClipBounds();
}

SKWASM_EXPORT void canvas_getDeviceClipBounds(Canvas* canvas, IRect* outRect) {
  *outRect = canvas->getDeviceClipBounds();
}
